internal/storage/v2/v1adapter/translator.go

```diff
@@ -63,6 +95,134 @@ func V1TracesFromSeq2(otelSeq iter.Seq2[[]ptrace.Traces, error]) ([]*model.Trace
 	return jaegerTraces, nil
 }
 
+// applyTraceSizeLimit applies trace size limits incrementally during sequence consumption.
+// This prevents loading large traces entirely into memory before applying the limit.
+func applyTraceSizeLimit(otelSeq iter.Seq2[[]ptrace.Traces, error], maxTraceSize int) iter.Seq2[[]ptrace.Traces, error] {
+	if maxTraceSize <= 0 {
+		return otelSeq
+	}
+
+	return func(yield func(traces []ptrace.Traces, err error) bool) {
+		otelSeq(func(traces []ptrace.Traces, err error) bool {
+			if err != nil {
+				return yield(traces, err)
+			}
+
+			// Pre-allocate slice with known capacity for better performance
+			limitedTraces := make([]ptrace.Traces, 0, len(traces))
+
+			// Process each trace in the batch and apply size limits incrementally
+			for _, trace := range traces {
+				limitedTrace := applyTraceSizeLimitToSingleTrace(trace, maxTraceSize)
+				limitedTraces = append(limitedTraces, limitedTrace)
+			}
+
+			return yield(limitedTraces, nil)
+		})
+	}
+}
+
+// applyTraceSizeLimitToSingleTrace applies size limit to a single trace with true incremental processing.
+// This version NEVER loads more spans than the limit, preventing OOM at the source.
+func applyTraceSizeLimitToSingleTrace(trace ptrace.Traces, maxTraceSize int) ptrace.Traces {
```

Comment:
I don't understand the signature of this function. One large trace can be split across many consecutive `ptrace.Traces` objects in a sequence. To apply a limit you need to accumulate them up to the limit and then skip the remaining ones.